# EDA

This notebook imports the data and explores the current format including visualizing the data.  

In [1]:
# imports 
import pandas as pd 
import numpy as np 
import re

In [2]:
import os 
import sys

module_path = os.path.abspath(os.path.join(os.pardir, os.pardir))
if module_path not in sys.path:
    sys.path.append(module_path)
    
from src import helpers as h

In [3]:
parent_dir = '../../'

In [4]:
df = pd.read_csv(parent_dir + 'data/weather_madrid_LEMD_1997_2015.csv', parse_dates = ['CET'])

In [5]:
df.head()

CET  Max TemperatureC  Mean TemperatureC  Min TemperatureC  \
0 1997-01-01               7.0                4.0               2.0   
1 1997-01-02               7.0                3.0               0.0   
2 1997-01-03               5.0                3.0               2.0   
3 1997-01-04               7.0                3.0              -1.0   
4 1997-01-05               2.0                0.0              -1.0   

   Dew PointC  MeanDew PointC  Min DewpointC  Max Humidity   Mean Humidity  \
0         5.0             3.0            2.0         100.0            95.0   
1         6.0             3.0            0.0         100.0            92.0   
2         5.0             1.0           -1.0         100.0            85.0   
3        -2.0            -3.0           -4.0          86.0            63.0   
4         2.0             0.0           -3.0         100.0            95.0   

    Min Humidity  ...   Max VisibilityKm   Mean VisibilityKm  \
0           76.0  ...               10.0                 9.0   
1           71.0  ...               10.0                 9.0   
2           70.0  ...               10.0                10.0   
3           49.0  ...               10.0                10.0   
4           86.0  ...               10.0                 5.0   

    Min VisibilitykM   Max Wind SpeedKm/h   Mean Wind SpeedKm/h  \
0                4.0                   13                     6   
1                4.0                   26                     8   
2                7.0                   27                    19   
3               10.0                   27                    19   
4                1.0                   14                     6   

    Max Gust SpeedKm/h  Precipitationmm   CloudCover     Events  \
0                  NaN              0.0          6.0        NaN   
1                 47.0              0.0          5.0       Rain   
2                  NaN              0.0          6.0  Rain-Snow   
3                 40.0              0.0          2.0        NaN   
4                  NaN              0.0          7.0       Snow   

   WindDirDegrees  
0             229  
1             143  
2             256  
3             284  
4               2  

[5 rows x 23 columns]

Check data types, number of entries, nulls etc:

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6812 entries, 0 to 6811
Data columns (total 23 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   CET                          6812 non-null   datetime64[ns]
 1   Max TemperatureC             6810 non-null   float64       
 2   Mean TemperatureC            6809 non-null   float64       
 3   Min TemperatureC             6810 non-null   float64       
 4   Dew PointC                   6810 non-null   float64       
 5   MeanDew PointC               6810 non-null   float64       
 6   Min DewpointC                6810 non-null   float64       
 7   Max Humidity                 6810 non-null   float64       
 8    Mean Humidity               6810 non-null   float64       
 9    Min Humidity                6810 non-null   float64       
 10   Max Sea Level PressurehPa   6812 non-null   int64         
 11   Mean Sea Level PressurehPa  6812 non-null 

All datatypes make sense so I don't think we need to make any adjustments right now.  

It looks like some of our column headers have preceding spaces which will get annoying.  There's also variability in how the column headers are named so I'm going to get all the column names in a more useable, pythonic form (all lowercase and underscores instead of spaces and no camel case):

In [7]:
df.columns = list(map(lambda x: h.make_pythonic(x), df.columns))

In [8]:
df.columns

Index(['cet', 'max_temperature_c', 'mean_temperature_c', 'min_temperature_c',
       'dew_point_c', 'mean_dew_point_c', 'min_dewpoint_c', 'max_humidity',
       'mean_humidity', 'min_humidity', 'max_sea_level_pressureh_pa',
       'mean_sea_level_pressureh_pa', 'min_sea_level_pressureh_pa',
       'max_visibility_km', 'mean_visibility_km', 'min_visibilityk_m',
       'max_wind_speed_km/h', 'mean_wind_speed_km/h', 'max_gust_speed_km/h',
       'precipitationmm', 'cloud_cover', 'events', 'wind_dir_degrees'],
      dtype='object')

Let me just manually change the date column and that one.........

In [9]:
df.rename(columns = {'min_visibilityk_m': 'min_visibility_km', 'cet': 'date'}, inplace = True);

In [10]:
df.columns

Index(['date', 'max_temperature_c', 'mean_temperature_c', 'min_temperature_c',
       'dew_point_c', 'mean_dew_point_c', 'min_dewpoint_c', 'max_humidity',
       'mean_humidity', 'min_humidity', 'max_sea_level_pressureh_pa',
       'mean_sea_level_pressureh_pa', 'min_sea_level_pressureh_pa',
       'max_visibility_km', 'mean_visibility_km', 'min_visibility_km',
       'max_wind_speed_km/h', 'mean_wind_speed_km/h', 'max_gust_speed_km/h',
       'precipitationmm', 'cloud_cover', 'events', 'wind_dir_degrees'],
      dtype='object')

Now, let's preview data again:

In [28]:
df.head()

date  max_temperature_c  mean_temperature_c  min_temperature_c  \
0 1997-01-01                7.0                 4.0                2.0   
1 1997-01-02                7.0                 3.0                0.0   
2 1997-01-03                5.0                 3.0                2.0   
3 1997-01-04                7.0                 3.0               -1.0   
4 1997-01-05                2.0                 0.0               -1.0   

   dew_point_c  mean_dew_point_c  min_dewpoint_c  max_humidity  mean_humidity  \
0          5.0               3.0             2.0         100.0           95.0   
1          6.0               3.0             0.0         100.0           92.0   
2          5.0               1.0            -1.0         100.0           85.0   
3         -2.0              -3.0            -4.0          86.0           63.0   
4          2.0               0.0            -3.0         100.0           95.0   

   min_humidity  ...  max_visibility_km  mean_visibility_km  \
0          76.0  ...               10.0                 9.0   
1          71.0  ...               10.0                 9.0   
2          70.0  ...               10.0                10.0   
3          49.0  ...               10.0                10.0   
4          86.0  ...               10.0                 5.0   

   min_visibility_km  max_wind_speed_km/h  mean_wind_speed_km/h  \
0                4.0                   13                     6   
1                4.0                   26                     8   
2                7.0                   27                    19   
3               10.0                   27                    19   
4                1.0                   14                     6   

   max_gust_speed_km/h  precipitationmm  cloud_cover     events  \
0                  NaN              0.0          6.0        NaN   
1                 47.0              0.0          5.0       Rain   
2                  NaN              0.0          6.0  Rain-Snow   
3                 40.0              0.0          2.0        NaN   
4                  NaN              0.0          7.0       Snow   

   wind_dir_degrees  
0               229  
1               143  
2               256  
3               284  
4                 2  

[5 rows x 23 columns]

**Let's see where most of the nan's are:**

In [16]:
df.isna().sum()

date                              0
max_temperature_c                 2
mean_temperature_c                3
min_temperature_c                 2
dew_point_c                       2
mean_dew_point_c                  2
min_dewpoint_c                    2
max_humidity                      2
mean_humidity                     2
min_humidity                      2
max_sea_level_pressureh_pa        0
mean_sea_level_pressureh_pa       0
min_sea_level_pressureh_pa        0
max_visibility_km               940
mean_visibility_km              940
min_visibility_km               940
max_wind_speed_km/h               0
mean_wind_speed_km/h              0
max_gust_speed_km/h            3306
precipitationmm                   0
cloud_cover                    1372
events                         5014
wind_dir_degrees                  0
dtype: int64

So out of 6812 entries, 5014 have missing events.  What are events:

In [25]:
df['events'].unique()

array([nan, 'Rain', 'Rain-Snow', 'Snow', 'Fog', 'Fog-Rain',
       'Rain-Thunderstorm', 'Thunderstorm', 'Rain-Hail-Thunderstorm',
       'Fog-Thunderstorm', 'Tornado', 'Fog-Rain-Thunderstorm',
       'Fog-Rain-Snow', 'Fog-Snow', 'Rain-Snow-Thunderstorm', 'Rain-Hail'],
      dtype=object)

So perhaps 'nan's just signify a 'normal' weather day - i.e. 'no event'.  Makes sense. 

**Check duplicates:**

In [24]:
df.duplicated().sum()

0

**Summary Stats:**

In [29]:
df.describe().T

count         mean         std    min     25%  \
max_temperature_c            6810.0    21.039648    8.867187    0.0    13.0   
mean_temperature_c           6809.0    14.658687    7.580461   -3.0     8.0   
min_temperature_c            6810.0     8.640529    6.837626  -10.0     3.0   
dew_point_c                  6810.0     8.120705    4.741067  -12.0     5.0   
mean_dew_point_c             6810.0     4.976211    4.654270  -15.0     2.0   
min_dewpoint_c               6810.0     1.451248    4.909705  -22.0    -2.0   
max_humidity                 6810.0    81.139354   17.531839   16.0    68.0   
mean_humidity                6810.0    57.971366   19.675744   15.0    41.0   
min_humidity                 6810.0    34.729369   19.320359    4.0    19.0   
max_sea_level_pressureh_pa   6812.0  1020.529360    6.235941  994.0  1017.0   
mean_sea_level_pressureh_pa  6812.0  1017.973136    6.480085  986.0  1014.0   
min_sea_level_pressureh_pa   6812.0  1015.217410    6.944745  965.0  1011.0   
max_visibility_km            5872.0    14.644074    8.770024    1.0    10.0   
mean_visibility_km           5872.0    11.719857    5.592324    0.0    10.0   
min_visibility_km            5872.0     9.134877    5.075065    0.0     7.0   
max_wind_speed_km/h          6812.0    21.953171    9.903914    0.0    14.0   
mean_wind_speed_km/h         6812.0     9.170728    5.110013    0.0     6.0   
max_gust_speed_km/h          3506.0    43.988306   12.252462   19.0    35.0   
precipitationmm              6812.0     0.111182    0.967174    0.0     0.0   
cloud_cover                  5440.0     3.206066    1.808948    0.0     2.0   
wind_dir_degrees             6812.0   197.234586  119.872777   -1.0    66.0   

                                50%      75%     max  
max_temperature_c              20.0    29.00    41.0  
mean_temperature_c             14.0    21.00    32.0  
min_temperature_c               9.0    14.00    28.0  
dew_point_c                     8.0    12.00    20.0  
mean_dew_point_c                6.0     8.00    16.0  
min_dewpoint_c                  2.0     5.00    14.0  
max_humidity                   87.0    94.00   100.0  
mean_humidity                  59.0    74.00   100.0  
min_humidity                   32.0    47.75   100.0  
max_sea_level_pressureh_pa   1020.0  1024.00  1047.0  
mean_sea_level_pressureh_pa  1018.0  1022.00  1043.0  
min_sea_level_pressureh_pa   1015.0  1019.00  1041.0  
max_visibility_km              10.0    10.00    31.0  
mean_visibility_km             10.0    10.00    31.0  
min_visibility_km              10.0    10.00    31.0  
max_wind_speed_km/h            21.0    27.00   182.0  
mean_wind_speed_km/h            8.0    11.00    39.0  
max_gust_speed_km/h            42.0    52.00   103.0  
precipitationmm                 0.0     0.00    32.0  
cloud_cover                     3.0     5.00     8.0  
wind_dir_degrees              223.0   299.00   360.0